source/val/validate_builtins.cpp

```diff
@@ -3218,16 +3240,31 @@ spv_result_t BuiltInsValidator::ValidateI32Vec4InputAtDefinition(
 
 spv_result_t BuiltInsValidator::ValidateWorkgroupSizeAtDefinition(
     const Decoration& decoration, const Instruction& inst) {
-  if (spv_result_t error = ValidateI32Vec(
-          decoration, inst, 3,
-          [this, &inst](const std::string& message) -> spv_result_t {
-            return _.diag(SPV_ERROR_INVALID_DATA, &inst)
-                   << _.VkErrorID(4427) << "According to the "
-                   << spvLogStringForEnv(_.context()->target_env)
-                   << " spec BuiltIn WorkgroupSize variable needs to be a "
-                      "3-component 32-bit int vector. "
-                   << message;
-          })) {
+  // Vulkan requires 32-bit int, but Universal can have 64-bit int
+  if (spvIsVulkanEnv(_.context()->target_env)) {
+    if (spv_result_t error = ValidateI32Vec(
+            decoration, inst, 3,
+            [this, &inst](const std::string& message) -> spv_result_t {
+              return _.diag(SPV_ERROR_INVALID_DATA, &inst)
+                     << _.VkErrorID(4427) << "According to the "
+                     << spvLogStringForEnv(_.context()->target_env)
+                     << " spec BuiltIn WorkgroupSize variable needs to be a "
+                        "3-component 32-bit int vector. "
+                     << message;
+            })) {
+      return error;
+    }
+  } else if (spv_result_t error = ValidateIVec(
+                 decoration, inst, 3,
+                 [this, &inst](const std::string& message) -> spv_result_t {
+                   return _.diag(SPV_ERROR_INVALID_DATA, &inst)
+                          << "According to the "
+                          << spvLogStringForEnv(_.context()->target_env)
+                          << " spec BuiltIn WorkgroupSize variable needs to be "
+                             "a "
+                             "3-component int vector. "
+                          << message;
+                 })) {
```

Comment:
I don't believe the universal validation rules say it has to be a vector.

The description of the workgroup in the spir-v spec says:

> Workgroup: The set of invocations partitioned in some execution models (e.g. GLCompute, Kernel) as a workgroup. Its size is defined statically by either the WorkgroupSize built-in or the LocalSize or LocalSizeId Execution Modes, or can be queried via the WorkgroupSize built-in. These values can be defined in multiple dimensions, and its total size is the product of the size in each specified dimension.

Note that is says it "can be defined in multiple dimensions." If OpenCL says it has to be a vector, then you can check that the environment is OpenCL.

Reply:
ok, will just remove the check all together for non-Vulkan